drivers/usb/device/usb_dc_stm.c

```diff
@@ -0,0 +1,586 @@
+
+	SYS_LOG_DBG("ep 0x%02x", ep);
+
+	if (!ep_state)
+		return -EINVAL;
+
+	*stalled = ep_state->ep_stalled;
+
+	return 0;
+}
+
+int usb_dc_ep_enable(const u8_t ep)
+{
+	struct usb_dc_stm_ep_state *ep_state = usb_dc_stm_get_ep_state(ep);
+	HAL_StatusTypeDef status;
+	int ret;
+
+	SYS_LOG_DBG("ep 0x%02x", ep);
+
+	if (!ep_state)
+		return -EINVAL;
+
+	SYS_LOG_DBG("HAL_PCD_EP_Open(0x%02x, %u, %u)", ep,
+		    ep_state->ep_mps, ep_state->ep_type);
+
+	status = HAL_PCD_EP_Open(&usb_dc_stm_state.pcd, ep,
+				 ep_state->ep_mps, ep_state->ep_type);
+	if (status != HAL_OK) {
+		SYS_LOG_ERR("HAL_PCD_EP_Open failed(0x%02x), %d",
+			    ep, (int)status);
+		return -EIO;
+	}
+
+	ret = usb_dc_ep_clear_stall(ep);
+	if (ret)
+		return ret;
+
+	if (EP_IS_OUT(ep) && ep != EP0_OUT) {
+		ret = usb_dc_ep_start_read(ep,
+					   usb_dc_stm_state.ep_buf[EP_IDX(ep)],
+					   USB_OTG_FS_MAX_PACKET_SIZE);
+		if (ret)
+			return ret;
+	}
+
+	return 0;
+}
+
+int usb_dc_ep_disable(const u8_t ep)
+{
+	struct usb_dc_stm_ep_state *ep_state = usb_dc_stm_get_ep_state(ep);
+	HAL_StatusTypeDef status;
+
+	SYS_LOG_DBG("ep 0x%02x", ep);
+
+	if (!ep_state)
+		return -EINVAL;
+
+	status = HAL_PCD_EP_Close(&usb_dc_stm_state.pcd, ep);
+	if (status != HAL_OK) {
+		SYS_LOG_ERR("HAL_PCD_EP_Close failed(0x%02x), %d",
+			    ep, (int)status);
+		return -EIO;
+	}
+
+	return 0;
+}
+
+int usb_dc_ep_write(const u8_t ep, const u8_t *const data,
+		    const u32_t data_len, u32_t * const ret_bytes)
+{
+	HAL_StatusTypeDef status;
+	int ret = 0;
+
+	SYS_LOG_DBG("ep 0x%02x, len %u", ep, data_len);
+
+	if (!EP_IS_IN(ep)) {
+		SYS_LOG_ERR("invalid ep 0x%02x", ep);
+		return -EINVAL;
+	}
+
+	if (!k_is_in_isr())
+		irq_disable(STM32F4_IRQ_OTG_FS);
+
+	status = HAL_PCD_EP_Transmit(&usb_dc_stm_state.pcd, ep,
+				     (void *)data, data_len);
+	if (status != HAL_OK) {
+		SYS_LOG_ERR("HAL_PCD_EP_Transmit failed(0x%02x), %d",
+			    ep, (int)status);
+		ret = -EIO;
+	}
+
+	if (!ret && ep == EP0_IN) {
+		/* Wait for an empty package as from the host.
+		 * This also flushes the TX FIFO to the host.
+		 */
+		usb_dc_ep_start_read(ep, NULL, 0);
+	}
+
+	if (!k_is_in_isr())
+		irq_enable(STM32F4_IRQ_OTG_FS);
+
+	*ret_bytes = data_len;
+
+	return ret;
+}
+
+int usb_dc_ep_read(const u8_t ep, u8_t *const data,
+		   const u32_t max_data_len, u32_t * const read_bytes)
+{
+	struct usb_dc_stm_ep_state *ep_state = usb_dc_stm_get_ep_state(ep);
+	u32_t read_count = ep_state->read_count;
+
+	SYS_LOG_DBG("ep 0x%02x, %u bytes, %u+%u, %p", ep,
+		    max_data_len, ep_state->read_offset, read_count, data);
+
+	if (max_data_len) {
+		if (read_count > max_data_len)
+			read_count = max_data_len;
+
+		if (read_count) {
+			memcpy(data,
+			       usb_dc_stm_state.ep_buf[EP_IDX(ep)] +
+			       ep_state->read_offset, read_count);
+			ep_state->read_count -= read_count;
+			ep_state->read_offset += read_count;
+		}
+
+		if (ep != EP0_OUT && !ep_state->read_count) {
+			usb_dc_ep_start_read(ep,
```

Comment:
It's still confusing, I guess comment at L:32..L:39 explaining it?

Reply:
Yes I'm going to add some comments.

Reply:
done